ls1intum/Artemis
# `Development`: Add option to disable lecture & quiz subsystems

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
In the future, we want to support deployments where not all instances of Artemis support all features and some features might not be available on some deployments at all.

This PR adds an option to
- disable the lecture-related endpoints on certain instances (Proof of Concept) and
- disable quiz-related logic on certain instances (also PoC)
- disable routes and buttons based on available profiles

### Description
**Note:** To avoid breaking changes, the Lecture/Quiz components are currently only disabled if the `decoupling` profile is set and the `lecture`/`quiz` profile is not set. This ensures that no breaking changes are introduced since deployments without the `decoupling` profile set will not have the endpoints disabled, even if the `lecture`/`quiz` profile is not set.


We want to be able to scale certain subcomponents of Artemis independent of each other, e.g. to scale up highly used subcomponents or ensuring that other parts of Artemis are still available, even if one subcomponent is faulty.

This PR disables the lecture-related endpoints, if the `lecture` profile is not activated (see Note above for more details); lecture-related requests are only sent to instances with the profile enabled.
The same is applied to quiz-related logic.
This requires additional config changes (if using a deployment using a manually-managed nginx proxy server) or a deployment via Kubernetes, which can automatically forward the requests to the correct instance.

As this PR allows to disable certain parts of Artemis in the server, these disabled parts should not be availabe in the client as otherwise, errors **will** occur.
We do this using the following approach:
- Each Artemis instance shares their available profiles using Hazelcast attributes
- Whenever an instance is started or stopped, all instances are notified and recalculate the union of the available profiles
- A WebSocket-subscription is established that broadcasts this union to all connected clients
- Clients will disable certain parts of the application (by disabling routes and buttons/links) based on these WebSocket messages (similar to the Feature-toggles)

Additional information for quiz-related changes:
- Block access to quiz exercises if not available on the system
- Only handle Quiz logic by instances that run the quiz profile
  - REST-calls are routed by nginx to the correct instances
  - WebSocket messages are either handled directly if the user is connected to a instance that runs quiz logic or forwarded to another instance using JMS-messages.
    - The JMS messages are distributed by the broker (ActiveMQ in our case) in a queue-logic so that the requests are only handled once.
    - One instance will process the message and act accordingly (by e.g. saving the submission to the database).
    - Other instances will not handle the message


### Steps for Testing
This PR requires config/deployment changes and can only be tested locally as of now.

Prerequisites:
- Local Artemis installation with
  - registry
  - at least 2 instances

1. Make sure the registry is running
2. Start one instance of Artemis with the `decoupling` profile, but without the `lecture` profile
3. Connect to this instance using your browser: Lecture-related buttons should be disabled
![image](https://user-images.githubusercontent.com/5084100/201037319-e2240311-5094-4699-8a8f-f016c7474e12.png)

4. Start the second instance with the `decoupling` and `lecture` profiles
5. After the instance is started, the client opened in step 3 should show the lecture-related aspects as enabled, no reload should be required
![image](https://user-images.githubusercontent.com/5084100/201037779-9eacc4a0-9ae8-46ec-8081-5037b9e303a2.png)
6. Ensure quiz functionality is still working, both in a setup with multiple instances and single-node environments

### Review Progress





#### Code Review
#### Manual Tests
